### DISCUSSION TITLE
[RFC] Overridable esbuild configs

### DISCUSSION THREAD

**Discussion by aiji42** [remix-run/remix#2153] [Proposals]
Currently (v1.2.2), the Remix compiler uses esbuild for both the browser and server compilers.
Depending on the values of `serverModuleFormat` and `serverBuildTarget` specified in remix.config.js, you can slightly control the configuration of esbuild, but you cannot change the configuration flexibly such as adding plugins.

Next.js uses webpack, and it is possible to change webpack settings from next.config.js. [Custom Webpack Config](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config)
The webpack function takes the original configuration value as the first argument and the context (server/client decision, webpack version information, etc.) as the second argument, and the configuration value can be overridden through the function.

Remix should also provide a function in remix.config.js that can override the esbuild configuration values, just like Next.js.

For example, something like this.
```js
import alias from "esbuild-plugin-alias"

module.exports = {
  // ...
  esbuild: (config, context) => {
    if (context.isServer) {
      config.plugins = [
        ...config.plugins,
        alias({ "some-module": "replace-some-module" })
      ]
    }
    config.jsxFactory = "jsx"
    return config
  },
}
```

If there is a lot of support for this proposal, I would like to start contributing this.
Please let me know what you think. Also, if you have similar issues and discussions, please let me know in the comments.

**Comment by MichaelDeBoey** [comment]
As I mentioned in your PR (https://github.com/remix-run/remix/pull/2168#issuecomment-1056026749), it's intentional that the `esbuild` config isn't exposed, so that the project isn't tied to `esbuild` and can be changed to something else whenever the team wants to (they already changed from `rollup` to `esbuild` once).

**Comment by MichaelDeBoey** [comment]
@kentcdodds's message about this in https://github.com/remix-run/remix/pull/2168#issuecomment-1058193715

> Thanks for putting in the effort @aiji42. This is extremely unlikely to happen. I'm not sure whether @ryanflorence or @mjackson have written out why we don't expose the build config anywhere (I think they have, but can't remember where), but I can say briefly that by exposing the build config it makes it too easy for people to shoot themselves in the foot and limits our ability to make changes that don't break people's apps.
> 
> We're much more interested in helping address specific use cases that people have and building in solutions for common use cases to building web apps.
> 
> For the edge cases, I know it's probably not what you want to hear, but in the early days of Remix, most of the Remix community made use of https://npm.im/patch-package to get features that Remix hadn't yet implemented and it actually worked out great. It's not something we'd necessarily recommend, but it's a pretty solid solution for folks who need something that we're planning on supporting in the future but isn't ready yet.

**Reply by mayank99** [reply]
Is there an example somewhere of using `patch-package` to use CSS-in-JS solutions like `linaria` or `vanilla-extract` (both of which provide an esbuild plugin)? Maybe someone from the "early days" was able to override the esbuild config to add a plugin?
